bitcoin/src/blockdata/transaction.rs

```diff
@@ -526,6 +534,20 @@ impl TxOut {
         Weight::from_vb(self.size() as u64).expect("should never happen under normal conditions")
     }
 
+    /// Returns the effective_value `Amount` returning `None` on overflow.
+    ///
+    /// The effective value is the value of this output minus the amount to spend it.  That is,
+    /// the effective_value can be calculated as value - (fee_rate * weight).
```

Comment:
Should have a warning about too many inputs.

Reply:
I'm actually not sure that adding more outputs changes the `effective_value` of a given utxo, because I believe that has no impact on the satisfaction_weight nor the fee_rate.  Adding more outputs changes the size of the transaction when you have more than 252 from what I've read, but should this actually be stated here?

Reply:
The point of `effective_value` used in coin-selection is to de-couple the fee computation of the inputs from the fee computation of the non_input_weight.  In your example above, if there is a solution using more than 252 inputs, then it's not the effective_value of the utxos that needs to be re-considered, but instead the `fee_rate * non_input_weight` that needs to be computed again.  In other-words, this is something that should be part coin-selection, but I don't believe has an impact on the `effective_value` calculation, since the effective_value is anchored on the size of utxo and the fee which does not change if more than 252 or less than 252 utxos are used.

Reply:
> the effective_value calculation, since the effective_value is anchored on the size of utxo and the fee which does not change if more than 252 or less than 252 utxos are used.

edit: the effective_value calculation, since the effective_value is anchored on the size of utxo and the feerate which does not change if more than 252 or less than 252 utxos are used.

Reply:
Sorry to keep messaging about this.  I think I was wrong earlier.  Does a transaction with more than 252 utxos have a different utxo weight for each utxo?  I guess I could also post and ask on bitcoin stack exchange also.

Reply:
> In other-words, this is something that should be part coin-selection, but I don't believe has an impact on the `effective_value` calculation, since the effective_value is anchored on the size of utxo and the fee which does not change if more than 252 or less than 252 utxos are used.

Yes, this is correct, it's just subtlety people may forget about so it's nice to warn them in the doc. To put it differently, having this nice API may lead people to think "writing coin selection algo is easy, just call this method and sum the inputs".

In general, since this is a high-stakes software, our approach is to try making the API impossible to misuse but if we can't we write appropriate warnings about edge cases in the docs. Even if the issues are not directly related.

Reply:
What about putting a description of the state machine on transaction about building the transaction with `n` outputs?  It sounds like it doesn't directly affect `effective_value`, although I think there should be a warning with a link to the aforementioned state machine.

Reply:
I don't know which state machine you're talking about.

Reply:
> I don't know which state machine you're talking about.

There isn't one yet.  I'm proposing adding a diagram of the state to help explain the impact of creating a transaction with a different number of utxos.  I think this "state machine" could be added to the documentation of `Transaction` and then we link to it from here and other functions/methods where we want to warn about blindly summing utxos..